channel/funder.go

```diff
@@ -0,0 +1,231 @@
+
+polling:
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(f.pollingInterval):
+			log.Println("Party B: Polling for opened channel...")
+			chanState, err := f.GetChannelState(ctx, req.Params, req.State)
+			if err != nil {
+				log.Println("Party B: Error while polling for opened channel:", err)
+				continue polling
+			}
+			log.Println("Party B: Found opened channel!")
+			if chanState.Control.FundedA && chanState.Control.FundedB {
+				return nil
+			}
+			return f.FundChannel(ctx, req.Params, req.State, true)
+		}
+	}
+}
+
+func (f *Funder) OpenChannel(ctx context.Context, params *pchannel.Params, state *pchannel.State) error {
+
+	perunAddress := f.GetPerunID()
+	kp := f.kpFull
+
+	openTxArgs := env.BuildOpenTxArgs(params, state)
+	txMeta, err := f.stellarClient.InvokeAndProcessHostFunction("open", openTxArgs, perunAddress, kp)
+	if err != nil {
+		return errors.New("error while invoking and processing host function: open")
+	}
+
+	_, err = DecodeEventsPerun(txMeta)
+	if err != nil {
+		return errors.New("error while decoding events")
+	}
+
+	return nil
+}
+
+func (f *Funder) FundChannel(ctx context.Context, params *pchannel.Params, state *pchannel.State, funderIdx bool) error {
+
+	perunAddress := f.GetPerunID()
+	tokenAddress := f.GetAssetID()
+
+	kp := f.kpFull
+	chanId := state.ID
+
+	fundTxArgs, err := env.BuildFundTxArgs(chanId, funderIdx)
+	if err != nil {
+		return errors.New("error while building fund tx")
+	}
+	balsStellar, err := wire.MakeBalances(state.Allocation)
+	if err != nil {
+		return errors.New("error while making balances")
+	}
+
+	tokenIDAddrFromBals := balsStellar.Token
+
+	sameContractTokenID := tokenIDAddrFromBals.Equals(tokenAddress)
+	if !sameContractTokenID {
+		panic("tokenIDAddrFromBals not equal to tokenContractAddress")
+	}
+
+	txMeta, err := f.stellarClient.InvokeAndProcessHostFunction("fund", fundTxArgs, perunAddress, kp)
+	if err != nil {
+		return errors.New("error while invoking and processing host function: fund")
+	}
+
+	_, err = DecodeEventsPerun(txMeta)
+	if err != nil {
+		return errors.New("error while decoding events")
+	}
+
+	return nil
+}
+
+func (f *Funder) AbortChannel(ctx context.Context, params *pchannel.Params, state *pchannel.State) error {
+
+	contractAddress := f.GetPerunID()
+	kp := f.kpFull
+	chanId := state.ID
+
+	openTxArgs, err := env.BuildGetChannelTxArgs(chanId)
+	if err != nil {
+		return errors.New("error while building get_channel tx")
+	}
+	txMeta, err := f.stellarClient.InvokeAndProcessHostFunction("abort_funding", openTxArgs, contractAddress, kp)
+	if err != nil {
+		return errors.New("error while invoking and processing host function: abort_funding")
+	}
+
+	_, err = DecodeEventsPerun(txMeta)
+	if err != nil {
+		return errors.New("error while decoding events")
+	}
+
+	return nil
+}
+
+func (f *Funder) GetChannelState(ctx context.Context, params *pchannel.Params, state *pchannel.State) (wire.Channel, error) {
```

Comment:
This is again duplicate with the adjudicators and adjudicator-subs `GetChannelState`. It should also be part of the `client` package I was talking about. Same goes for `OpenChannel` and `FundChannel`.